driver/controller.go

```diff
@@ -802,44 +804,97 @@ func (d *Driver) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsReques
-		// index 31 up to 50.
-
-		var nextToken int
-		var err error
+		// Paginate through snapshots and return results.
+
+		// Pagination is controlled by two request parameters:
+		// MaxEntries indicates how many entries should be returned at most. If
+		// more results are available, we must return a NextToken value
+		// indicating the index for the next snapshot to request.
+		// StartingToken defines the index of the first snapshot to return.
+		// The CSI request parameters are defined in terms of number of
+		// snapshots, not pages. It is up to the driver to translate the
+		// parameters into paged requests accordingly.
+
+		var (
+			startingToken         int32
+			originalStartingToken int32
+		)
 		if req.StartingToken != "" {
-			nextToken, err = strconv.Atoi(req.StartingToken)
+			parsedToken, err := strconv.ParseInt(req.StartingToken, 10, 32)
 			if err != nil {
-				return nil, status.Errorf(codes.Aborted, "ListSnapshots starting token %s is not valid : %s",
-					req.StartingToken, err.Error())
+				return nil, status.Errorf(codes.Aborted, "ListSnapshots starting token %q is not valid: %s", req.StartingToken, err)
 			}
+			startingToken = int32(parsedToken)
+			originalStartingToken = startingToken
 		}
 
-		if nextToken != 0 && req.MaxEntries != 0 {
-			return nil, status.Errorf(codes.Aborted,
-				"ListSnapshots invalid arguments starting token: %d and max entries: %d can't be non null at the same time", nextToken, req.MaxEntries)
-		}
-
-		// fetch all entries
+		// Fetch snapshots until we have either collected req.MaxEntries (if
+		// positive) or all available ones, whichever comes first.
 		listOpts := &godo.ListOptions{
+			Page:    1,
 			PerPage: int(req.MaxEntries),
 		}
-		var snapshots []godo.Snapshot
+		if req.MaxEntries > 0 {
+			// MaxEntries also defines the page size so that we can skip over
+			// snapshots before the StartingToken and minimize the number of
+			// paged requests we need.
+			listOpts.Page = int(startingToken/req.MaxEntries) + 1
+			// Offset StartingToken to skip snapshots we do not want. This is
+			// needed when MaxEntries does not divide StartingToken without
+			// remainder.
+			startingToken = startingToken % req.MaxEntries
+		}
+
+		log = log.WithFields(logrus.Fields{
+			"page":                    listOpts.Page,
+			"computed_starting_token": startingToken,
+		})
+
+		var (
+			// remainingEntries keeps track of how much room is left to return
+			// as many as MaxEntries snapshots.
+			remainingEntries int = int(req.MaxEntries)
+			// hasMore indicates if NextToken must be set.
+			hasMore   bool
+			snapshots []godo.Snapshot
+		)
 		for {
+			hasMore = false
 			snaps, resp, err := d.snapshots.ListVolume(ctx, listOpts)
 			if err != nil {
-				return nil, status.Errorf(codes.Aborted, "ListSnapshots listing volume snapshots has failed: %s", err.Error())
+				return nil, status.Errorf(codes.Internal, "ListSnapshots listing volume snapshots has failed: %s", err)
```

Comment:
This status code was actually correct before: https://github.com/container-storage-interface/spec/blob/master/spec.md#listsnapshots-errors